ui/pages/routes/confirmation-handler.tsx

```diff
@@ -0,0 +1,156 @@
+import { getConfirmationRoute } from '../confirmations/hooks/useConfirmationNavigation';
+// eslint-disable-next-line import/no-restricted-paths
+import { getEnvironmentType } from '../../../app/scripts/lib/util';
+import {
+  ENVIRONMENT_TYPE_NOTIFICATION,
+  ENVIRONMENT_TYPE_FULLSCREEN,
+  SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES,
+} from '../../../shared/constants/app';
+import {
+  getHasApprovalFlows,
+  getHasBridgeQuotes,
+  selectPendingApprovalsForNavigation,
+} from '../../selectors';
+import {
+  getFetchParams,
+  getHasSwapsQuotes,
+  getShowAwaitingSwapScreen,
+} from '../../ducks/swaps/swaps';
+import { useNavState } from '../../contexts/navigation-state';
+
+const snapApprovals = [
+  'wallet_installSnap',
+  'wallet_updateSnap',
+  'wallet_installSnapResult',
+  ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
+  SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.confirmAccountCreation,
+  SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.confirmAccountRemoval,
+  SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.showNameSnapAccount,
+  SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.showSnapAccountRedirect,
+  ///: END:ONLY_INCLUDE_IF
+];
+
+export const ConfirmationHandler = () => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const { pathname } = location;
+  const navState = useNavState();
+
+  const envType = getEnvironmentType();
+  const isNotification = envType === ENVIRONMENT_TYPE_NOTIFICATION;
+  const isFullscreen = envType === ENVIRONMENT_TYPE_FULLSCREEN;
+
+  // In test environments (Jest integration tests), run synchronously
+  // In real browser, defer to avoid race conditions with Redux state updates
+  const isTestEnvironment = Boolean(process.env.IN_TEST);
+
+  const showAwaitingSwapScreen = useSelector(getShowAwaitingSwapScreen);
+  const hasSwapsQuotes = useSelector(getHasSwapsQuotes);
+  const hasBridgeQuotes = useSelector(getHasBridgeQuotes);
+  const swapsFetchParams = useSelector(getFetchParams);
+  const pendingApprovals = useSelector(selectPendingApprovalsForNavigation);
+  const hasApprovalFlows = useSelector(getHasApprovalFlows);
+
+  // Read stayOnHomePage from both v5 location.state and v5-compat navState
+  const stayOnHomePage = useMemo(
+    () =>
+      Boolean(location.state?.stayOnHomePage) ||
+      Boolean(navState?.stayOnHomePage),
+    [location.state, navState],
+  );
+
+  const canRedirect = !isNotification && !stayOnHomePage;
+
+  // Flows that *should* navigate in fullscreen, based on E2E specs
+  const hasSnapApproval = pendingApprovals.some((approval) =>
+    snapApprovals.includes(approval.type),
+  );
+
+  // Flows that *should not* navigate in fullscreen, based on E2E specs
+  const hasSmartTransactionStatus = pendingApprovals.some(
+    (approval) =>
+      approval.type === 'smartTransaction:showSmartTransactionStatusPage' &&
+      approval.origin !== 'metamask' &&
+      approval.origin !== 'MetaMask',
+  );
+
+  const skipHandler = useMemo(() => {
+    return (
+      isFullscreen &&
```

Comment:
Will this prevent navigating in full screen completely unless we have a snap approval or approval flows? Didn't we previously navigate in full screen from home as normal for transactions and signatures for example?